README.md
<a name="readme-top"></a>

[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]
[![LinkedIn][linkedin-shield]][linkedin-url]

<!-- PROJECT LOGO -->
<br />
<div align="center">
  <a href="https://github.com/AaravMalani/ConstantPool-Editor">
    <h1 align="center">Constant Pool Editor</h1>
  </a>



  <p align="center">
    🍵 A script to view and edit the constant pool of a Java .class file
    <br />
    <a href="https://github.com/AaravMalani/ConstantPool-Editor"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="#usage">View Demo</a>
    ·
    <a href="https://github.com/AaravMalani/ConstantPool-Editor/issues">Report Bug</a>
    ·
    <a href="https://github.com/AaravMalani/ConstantPool-Editor/issues">Request Feature</a>
  </p>
</div>



<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <ol>
    <li>
        <a href="#built-with">Built With</a>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#prerequisites">Prerequisites</a></li>
        <li><a href="#installation">Installation</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#contributing">Contributing</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#acknowledgments">Acknowledgments</a></li>
  </ol>
</details>


### Built With
[![Python][python-shield]][python-url]

<p align="right">(<a href="#readme-top">back to top</a>)</p>

<!-- GETTING STARTED -->
## Getting Started

### Prerequisites

* Python 3
  ```sh
  sudo apt install python3
  ```
  or
  ```sh
  sudo yum install python3
  ```
  or download from the [Python website](https://www.python.org/downloads/)

### Installation

  Clone the repo and you should be done!
   ```sh
   git clone https://github.com/AaravMalani/ConstantPool-Editor.git
   ```

<p align="right">(<a href="#readme-top">back to top</a>)</p>



<!-- USAGE EXAMPLES -->
## Usage

To run the code, use the following command
```sh
python3 main.py [options] <filename>
```

A full list of the options are given in the Wiki!

<p align="right">(<a href="#readme-top">back to top</a>)</p>



<!-- ROADMAP -->
## Roadmap

- [x] Edit and view the constant pool
- [x] Parse options using `argparse`
- [ ] Add elements to constant pool
- [ ] Add a GUI
- [ ] Add documentation  

See the [open issues](https://github.com/AaravMalani/ConstantPool-Editor/issues) for a full list of proposed features (and known issues).

<p align="right">(<a href="#readme-top">back to top</a>)</p>


<!-- CONTRIBUTING -->
## Contributing

Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any contributions you make are **greatly appreciated**.

If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag "enhancement".
Don't forget to give the project a star! Thanks again!

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request

<p align="right">(<a href="#readme-top">back to top</a>)</p>



<!-- LICENSE -->
## License

Distributed under the MIT License. See `LICENSE` for more information.

<p align="right">(<a href="#readme-top">back to top</a>)</p>


[contributors-shield]: https://img.shields.io/github/contributors/AaravMalani/ConstantPool-Editor.svg?style=for-the-badge
[contributors-url]: https://github.com/AaravMalani/ConstantPool-Editor/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/AaravMalani/ConstantPool-Editor.svg?style=for-the-badge
[forks-url]: https://github.com/AaravMalani/ConstantPool-Editor/network/members
[stars-shield]: https://img.shields.io/github/stars/AaravMalani/ConstantPool-Editor.svg?style=for-the-badge
[stars-url]: https://github.com/AaravMalani/ConstantPool-Editor/stargazers
[issues-shield]: https://img.shields.io/github/issues/AaravMalani/ConstantPool-Editor.svg?style=for-the-badge
[issues-url]: https://github.com/AaravMalani/ConstantPool-Editor/issues
[license-shield]: https://img.shields.io/github/license/AaravMalani/ConstantPool-Editor.svg?style=for-the-badge
[license-url]: https://github.com/AaravMalani/ConstantPool-Editor/blob/main/LICENSE
[linkedin-shield]: https://img.shields.io/badge/-LinkedIn-black.svg?style=for-the-badge&logo=linkedin&colorB=555
[linkedin-url]: https://linkedin.com/in/aaravm
[python-shield]: https://img.shields.io/badge/Python-333?style=for-the-badge&logo=python
[python-url]: https://python.org/ 


CHANGELOG
# Changelog
## Version 1.0.1
- Formatted some of the variable names for readability

javaparser.py
import pprint
import sys
import struct
import argparse
import os
__version__ = "1.0.0"
__author__ = 'Aarav Malani'
__license__ = 'MIT'

if os.name == 'nt':
    os.system("color")

parser = argparse.ArgumentParser(
    prog='ConstantPoolEditor',
    description='Edits the constant pool of a Java file')
parser.add_argument(
    "filename", help="The name of the .class file to parse", nargs='?')
parser.add_argument(
    "-v", "--version", help="Output the version of the program", action='store_true')
parser.add_argument(
    "-e", "--edit", help="Edit the constant pool (default is viewing only)", action='store_true')
parser.add_argument(
    "-r", "--resolve", help="Resolve the indexes in the constant pool", action='store_true')
parser.add_argument("-H", "--hide-tag",
                    help="Hides the tag and length of the constant pool elements", action='store_true')
args = parser.parse_args()
if args.version:
    print("ConstantPoolEditor "+__version__)
    with open('LICENSE') as f:
        print(f.read().replace(__license__+" License\n\n", ""))
    print("Written by "+__author__+".")
    sys.exit(0)
if not args.filename:
    print("ConstantPoolEditor: error: the following arguments are required: filename")
    sys.exit(1)


def nextBytes(x, d, func=int.from_bytes):
    return func(d[:x]), d[x:]


with open(args.filename, "rb") as f:
    b = f.read()
magic, b = nextBytes(4, b, bytes)
if magic != b'\xca\xfe\xba\xbe':
    print("ERROR: Invalid magic")
    sys.exit(0)
minor, b = nextBytes(2, b)
major, b = nextBytes(2, b)
a = {k: '1.'+str(i) for k, i in zip([45]+list(range(45, 64)), [*range(0, 20)])}
cpc, b = nextBytes(2, b)
cp = []


class CONSTANT:
    def __repr__(self):
        returned = "<"+self.__class__.__name__+" "
        for i in self.__dict__:
            if args.hide_tag and i in ['tag', 'length']:
                continue

            if args.resolve and i.endswith('_index'):
                try:
                    val = str(cp[self.__dict__[i]-1])
                    if type(val) is CONSTANT_Utf8:
                        val = val.bytes.decode('latin1')
                    returned += i+'='+val + ' '
                except:
                    returned += i+'='+str(self.__dict__[i])
            elif type(self.__dict__[i]) is bytes:
                returned += i + '='+self.__dict__[i].decode('latin1') + ' '
            else:
                returned += i + '='+str(self.__dict__[i]) + ' '
        returned = returned[:-1]
        returned += '>'

        return returned


def pack(dct):
    returned = b''
    for i in dct:
        key, value = tuple(list(i.items())[0])
        if type(key) is bytes:
            returned += key
        elif type(key) is int:
            returned += int.to_bytes(key, length=value)
        elif type(key) is float:
            returned += struct.pack(">f", key).rjust(value, '\x00')
    return returned


class CONSTANT_Utf8(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.length = int.from_bytes(data[1:3])
        self.bytes = data[3:]

    def pack(self):
        return pack([{self.tag: 1}, {self.length: 2}, {self.bytes: 0}])


class CONSTANT_Ref(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.class_index = int.from_bytes(data[1:3])
        self.name_and_type_index = int.from_bytes(data[3:])

    def pack(self):
        return pack([{self.tag: 1}, {self.class_index: 2}, {self.name_and_type_index: 2}])


class CONSTANT_Fieldref(CONSTANT_Ref):
    pass


class CONSTANT_Methodref(CONSTANT_Ref):
    pass


class CONSTANT_InterfaceMethodref(CONSTANT_Ref):
    pass


class CONSTANT_String(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.string_index = int.from_bytes(data[1:])

    def pack(self):
        return pack([{self.tag: 1}, {self.string_index: 2}])


class CONSTANT_Integer(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.bytes = int.from_bytes(data[1:5])

    def pack(self):
        return pack([{self.tag: 1}, {self.bytes: 4}])


class CONSTANT_Float(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.bytes = struct.unpack('>f', data[1:5])

    def pack(self):
        return pack([{self.tag: 1}, {self.bytes: 4}])


class CONSTANT_Long(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.high_bytes = data[1:5]
        self.low_bytes = data[5:9]

    def pack(self):
        return pack([{self.tag: 1}, {self.high_bytes: 4}, {self.low_bytes: 4}])


class CONSTANT_Double(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.high_bytes = data[1:5]
        self.low_bytes = data[5:9]

    def pack(self):
        return pack([{self.tag: 1}, {self.high_bytes: 4}, {self.low_bytes: 4}])


class CONSTANT_Class(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.name_index = int.from_bytes(data[1:])

    def pack(self):
        return pack([{self.tag: 1}, {self.name_index: 2}])


class CONSTANT_NameAndType(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.name_index = int.from_bytes(data[1:3])
        self.descriptor_index = int.from_bytes(data[3:])

    def pack(self):
        return pack([{self.tag: 1}, {self.name_index: 2}, {self.descriptor_index: 2}])


class CONSTANT_MethodHandle(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.reference_kind = int.from_bytes(data[1:2])
        self.reference_index = int.from_bytes(data[2:])

    def pack(self):
        return pack([{self.tag: 1}, {self.reference_kind: 1}, {self.reference_index: 2}])


class CONSTANT_MethodType(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.descriptor_index = int.from_bytes(data[1:])

    def pack(self):
        return pack([{self.tag: 1}, {self.descriptor_index: 2}])


class CONSTANT_InvokeDynamic(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.bootstrap_method_attr_index = int.from_bytes(data[1:3])
        self.name_and_type_index = int.from_bytes(data[3:])

    def pack(self):
        return pack([{self.tag: 1}, {self.bootstrap_method_attr_index: 2}, {self.name_and_type_index: 2}])


for i in range(cpc - 1):
    c, b = nextBytes(1, b)

    lengthTable = {
        7: 2,
        9: 4,
        10: 4,
        11: 4,
        8: 2,
        3: 4,
        4: 4,
        5: 8,
        6: 8,
        12: 4,
        1: 2,
        15: 3,
        16: 2,
        18: 4
    }
    try:
        data, b = nextBytes(lengthTable[c], b, func=bytes)
    except:
        print("Error loading file! Dumping constant pool!")
        pprint.pprint(cp)
        sys.exit(1)
    if c == 1:
        data2, b = nextBytes(int.from_bytes(data[:2]), b, bytes)
        data += data2
    classTable = {
        1: CONSTANT_Utf8,
        7: CONSTANT_Class,
        9: CONSTANT_Fieldref,
        10: CONSTANT_Methodref,
        11: CONSTANT_InterfaceMethodref,
        8: CONSTANT_String,
        3: CONSTANT_Integer,
        4: CONSTANT_Float,
        5: CONSTANT_Long,
        6: CONSTANT_Double,
        12: CONSTANT_NameAndType,
        15: CONSTANT_MethodHandle,
        16: CONSTANT_MethodType,
        18: CONSTANT_InvokeDynamic
    }
    cp += [classTable.get(c, bytes)(c.to_bytes(1)+data)]

print("Class File Version: "+a[major]+('.'+str(minor) if minor else ''))
print('Constant Pool Length: '+str(cpc))
for c, i in enumerate(cp):
    print(hex(c+1)[2:].zfill(4) + ': '+str(i))

if args.edit:
    while True:
        a = int(input("Index (Type -1 to save)? ")) - 1
        if a == -2:
            break
        c = list(enumerate(filter(lambda x: x[0] not in ['tag', 'length'], zip(
            cp[a].__dict__.keys(), cp[a].__dict__.values()))))
        for i in c:
            print(str(i[0] + 1) + '. '+str(i[1][0]) + ': '+str(i[1][1]))
        x = int(input("Choose value to edit? "))
        d = input('Enter value? ')

        os.system('cls' if os.name == 'nt' else 'clear')
        try:
            cp[a].__dict__[c[x-1][1][0]] = type(c[x-1][1][1])(d)
        except TypeError:
            cp[a].__dict__[c[x-1][1][0]
                           ] = type(c[x-1][1][1])(d, encoding='utf8')
        if not (cp[a].tag - 1):
            cp[a].length = len(cp[a].bytes)
        for c, i in enumerate(cp):
            print(hex(c+1)[2:].zfill(4) + ': '+str(i))
    with open(input("Save to [default is "+args.filename+"]? ") or args.filename, 'wb') as f:
        data = b'\xca\xfe\xba\xbe'+int.to_bytes(minor, 2)+int.to_bytes(
            major, 2)+int.to_bytes(cpc, 2) + b''.join([i.pack() for i in cp]) + b
        f.write(data)
        print("Saved and exiting!")


main.py
import argparse
import os
import pprint
import struct
import sys
from typing import Any

__version__ = "1.0.1"
__author__ = 'Aarav Malani'
__license__ = 'MIT'

if os.name == 'nt':
    os.system("color")

parser = argparse.ArgumentParser(
    prog='ConstantPoolEditor',
    description='Edits the constant pool of a Java file')


parser.add_argument(
    "filename", help="The name of the .class file to parse", nargs='?')
parser.add_argument(
    "-v", "--version", help="Output the version of the program", action='store_true')
parser.add_argument(
    "-e", "--edit", help="Edit the constant pool (default is viewing only)", action='store_true')
parser.add_argument(
    "-r", "--resolve", help="Resolve the indexes in the constant pool", action='store_true')
parser.add_argument("-H", "--hide-tag",
                    help="Hides the tag and length of the constant pool elements", action='store_true')

args = parser.parse_args()
if args.version:
    print("ConstantPoolEditor "+__version__)
    with open('LICENSE') as f:
        print(f.read().replace(__license__+" License\n\n", ""))
    print("Written by "+__author__+".")
    sys.exit(0)
if not args.filename:
    print("ConstantPoolEditor: error: the following arguments are required: filename")
    sys.exit(1)

def nextBytes(no_of_bytes: int, data: bytes, func: callable = int.from_bytes) -> tuple[Any,bytes]:
    """
    Splits the bytes at index `no_of_bytes` and returns a tupl
    For example
    ```
    nextBytes(2, b'data', func=bytes) ## (b'da',b'ta')
    ```
    
    Parameters:
    `no_of_bytes`: Index at which to split 
    `data`: The data to split
    `func`: The function to call on splitting (default is `int.from_bytes`)
    Returns:
    tuple: A two elemented tuple containing the slices of the bytes
    """
    return func(data[:no_of_bytes]), data[no_of_bytes:]


with open(args.filename, "rb") as f:
    classData : bytes = f.read()

magic, classData = nextBytes(4, classData, bytes)

if magic != b'\xca\xfe\xba\xbe': # All .class files start with cafebabe
    print("ERROR: Invalid magic", file=sys.stderr)
    sys.exit(1)

minor, classData = nextBytes(2, classData, bytes)
major, classData = nextBytes(2, classData, bytes)
major_int = int.from_bytes(major, byteorder='big')
if major_int not in range(45, 64):
    print("ERROR: Invalid major version! Perhaps try updating this software (Version "+__version__+")", file=sys.stderr)
    sys.exit(1)
constant_pool_count, classData = nextBytes(2, classData, bytes)

class CONSTANT:
    def __repr__(self):
        returned = "<"+self.__class__.__name__+" "
        for i in self.__dict__:
            if args.hide_tag and i in ['tag', 'length']:
                continue

            if args.resolve and i.endswith('_index'):
                try:
                    val = str(cp[self.__dict__[i]-1])
                    if type(val) is CONSTANT_Utf8:
                        val = val.bytes.decode('latin1')
                    returned += i+'='+val + ' '
                except:
                    returned += i+'='+str(self.__dict__[i])
            elif type(self.__dict__[i]) is bytes:
                returned += i + '='+self.__dict__[i].decode('latin1') + ' '
            else:
                returned += i + '='+str(self.__dict__[i]) + ' '
        returned = returned[:-1]
        returned += '>'

        return returned

cp : list[CONSTANT]= []

def pack(dct):
    returned = b''
    for i in dct:
        key, value = tuple(list(i.items())[0])
        if type(key) is bytes:
            returned += key
        elif type(key) is int:
            returned += int.to_bytes(key, length=value)
        elif type(key) is float:
            returned += struct.pack(">f", key).rjust(value, '\x00')
    return returned


class CONSTANT_Utf8(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.length = int.from_bytes(data[1:3])
        self.bytes = data[3:]

    def pack(self):
        return pack([{self.tag: 1}, {self.length: 2}, {self.bytes: 0}])


class CONSTANT_Ref(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.class_index = int.from_bytes(data[1:3])
        self.name_and_type_index = int.from_bytes(data[3:])

    def pack(self):
        return pack([{self.tag: 1}, {self.class_index: 2}, {self.name_and_type_index: 2}])


class CONSTANT_Fieldref(CONSTANT_Ref):
    pass


class CONSTANT_Methodref(CONSTANT_Ref):
    pass


class CONSTANT_InterfaceMethodref(CONSTANT_Ref):
    pass


class CONSTANT_String(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.string_index = int.from_bytes(data[1:])

    def pack(self):
        return pack([{self.tag: 1}, {self.string_index: 2}])


class CONSTANT_Integer(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.bytes = int.from_bytes(data[1:5])

    def pack(self):
        return pack([{self.tag: 1}, {self.bytes: 4}])


class CONSTANT_Float(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.bytes = struct.unpack('>f', data[1:5])

    def pack(self):
        return pack([{self.tag: 1}, {self.bytes: 4}])


class CONSTANT_Long(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.high_bytes = data[1:5]
        self.low_bytes = data[5:9]

    def pack(self):
        return pack([{self.tag: 1}, {self.high_bytes: 4}, {self.low_bytes: 4}])


class CONSTANT_Double(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.high_bytes = data[1:5]
        self.low_bytes = data[5:9]

    def pack(self):
        return pack([{self.tag: 1}, {self.high_bytes: 4}, {self.low_bytes: 4}])


class CONSTANT_Class(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.name_index = int.from_bytes(data[1:])

    def pack(self):
        return pack([{self.tag: 1}, {self.name_index: 2}])


class CONSTANT_NameAndType(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.name_index = int.from_bytes(data[1:3])
        self.descriptor_index = int.from_bytes(data[3:])

    def pack(self):
        return pack([{self.tag: 1}, {self.name_index: 2}, {self.descriptor_index: 2}])


class CONSTANT_MethodHandle(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.reference_kind = int.from_bytes(data[1:2])
        self.reference_index = int.from_bytes(data[2:])

    def pack(self):
        return pack([{self.tag: 1}, {self.reference_kind: 1}, {self.reference_index: 2}])


class CONSTANT_MethodType(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.descriptor_index = int.from_bytes(data[1:])

    def pack(self):
        return pack([{self.tag: 1}, {self.descriptor_index: 2}])


class CONSTANT_InvokeDynamic(CONSTANT):
    def __init__(self, data):
        self.tag = data[0]
        self.bootstrap_method_attr_index = int.from_bytes(data[1:3])
        self.name_and_type_index = int.from_bytes(data[3:])

    def pack(self):
        return pack([{self.tag: 1}, {self.bootstrap_method_attr_index: 2}, {self.name_and_type_index: 2}])


for i in range(len(constant_pool_count) - 1):
    tag, classData = nextBytes(1, classData, bytes)

    lengthTable = {
        7: 2,
        9: 4,
        10: 4,
        11: 4,
        8: 2,
        3: 4,
        4: 4,
        5: 8,
        6: 8,
        12: 4,
        1: 2,
        15: 3,
        16: 2,
        18: 4
    }
    try:
        data, classData = nextBytes(lengthTable[tag], classData, func=bytes)
    except:
        print("Error loading file! Dumping constant pool!")
        pprint.pprint(cp)
        sys.exit(1)
    if tag == 1: ## Special Utf-8 Case
        utfData, classData = nextBytes(int.from_bytes(data[:2]), classData, bytes)
        data += utfData
    classTable = {
        1: CONSTANT_Utf8,
        7: CONSTANT_Class,
        9: CONSTANT_Fieldref,
        10: CONSTANT_Methodref,
        11: CONSTANT_InterfaceMethodref,
        8: CONSTANT_String,
        3: CONSTANT_Integer,
        4: CONSTANT_Float,
        5: CONSTANT_Long,
        6: CONSTANT_Double,
        12: CONSTANT_NameAndType,
        15: CONSTANT_MethodHandle,
        16: CONSTANT_MethodType,
        18: CONSTANT_InvokeDynamic
    }
    cp += [classTable.get(tag, bytes)(tag.to_bytes(1)+data)]

print("Class File Version: "+major_to_string[major]+('.'+str(minor) if minor else ''))
print('Constant Pool Length: '+str(constant_pool_count))
for c, i in enumerate(cp):
    print(str(c+1).zfill(8) + ': '+str(i))

if args.edit:
    while True:
        try:
            index = int(input("Index (Enter nothing to save)? "))
        except:
            break
        values = list(filter(lambda x: x[0] not in ['tag', 'length'], zip(
            cp[index].__dict__.keys(), cp[index].__dict__.values())))
        for c, (k, v) in enumerate(values):
            print(str(c + 1) + '. '+str(k) + ': '+str(v))
        valIndex = int(input("Choose value to edit? "))
        value = input('Enter value? ')

        os.system('cls' if os.name == 'nt' else 'clear')
        try:
            cp[index].__dict__[c[valIndex-1][0]] = type(c[valIndex-1][1])(value)
        except TypeError: ## Pesky byte handling
            cp[index].__dict__[c[valIndex-1][0]] = type(c[valIndex-1][1])(value, encoding='utf8')
        if cp[index].tag == 1:
            cp[index].length = len(cp[index].bytes)
        for c, i in enumerate(cp):
            print(str(c+1).zfill(8) + ': '+str(i))
    with open(input("Save to [default is "+args.filename+"]? ") or args.filename, 'wb') as f:
        data = b'\xca\xfe\xba\xbe'+int.to_bytes(minor, 2)+int.to_bytes(
            major, 2)+int.to_bytes(constant_pool_count, 2) + b''.join([i.pack() for i in cp]) + classData
        f.write(data)
        print("Saved and exiting!")